src/test/util/net.h

```diff
@@ -30,4 +30,35 @@ struct ConnmanTestMsg : public CConnman {
     bool ReceiveMsgFrom(CNode& node, CSerializedNetMsg& ser_msg) const;
 };
 
+constexpr ServiceFlags ALL_SERVICE_FLAGS[]{
+    NODE_NONE,
+    NODE_NETWORK,
+    NODE_BLOOM,
+    NODE_WITNESS,
+    NODE_COMPACT_FILTERS,
+    NODE_NETWORK_LIMITED,
+};
```

Comment:
I wonder if it is possible to get a warning if a new enum value is added, but this array is forgotten to be updated? I think not :/

Reply:
It is ok as it is, and I am just wondering if this can be improved wrt minimizing the chance of the actual enum and the "all" arrays going out of sync over time. Maybe one of the below would help:

o Add a comment next to the enum definition "don't forget to update `ALL_...` in `src/test/util/net.h` if you change this enum".

o Define the array immediately below the enum itself.

<details>
<summary>o Define both in one struct</summary>

```cpp
struct Color
{
    enum E : uint8_t {
        RED,
        GREEN,
        BLUE,
    };

    static constexpr E all[]{
        RED,
        GREEN,
        BLUE,
    };
};

...
    std::cout << Color::BLUE << std::endl;

    for (auto& c : Color::all) {
        std::cout << c << std::endl;
    }
```
</details>

Reply:
Perhaps we could start using [`kMaxValue` as in the Chromium project](https://www.chromium.org/developers/coding-style/chromium-style-checker-errors). That way we could ​assert that the last element of the `ALL_*` array is equal to the `enum`'s `kMaxValue`.

Another benefit from using `kMaxValue` is that we could start using `FuzzedDataProvider::ConsumeEnum` when fuzzing.

Reply:
Yeah, that is another way. `enum Network` uses this. It only works as long as enum values are left at their defaults:

https://github.com/bitcoin/bitcoin/blob/f1f26b8d5baec4a45a3a9ba0440cd4eff7af8407/src/netaddress.h#L38

and also the dummy "max" value has to be explicitly handled in `switch` without `default:`

https://github.com/bitcoin/bitcoin/blob/f1f26b8d5baec4a45a3a9ba0440cd4eff7af8407/src/netaddress.cpp#L145-L146

Reply:
For strong enums, I was thinking about a macro `ENUM_CLASS_ALL` that writes the `enum class` and `std::array` that holds all enum types of that class. Thus, we could keep using `PickValueInArray` for strong enums. `FuzzedDataProvider::ConsumeEnum` is an alternative for strong enums that does exactly the same from the perspective of the fuzz engine.

For weak enums, `ConsumeWeakEnum` is probably the best bet. I tried `kMaxValue` and it performed strictly worse in all scenarios, in one it was unable to find a crash at all. And it can't possibly work for 64-bit enums. Someone should report those bugs to google. See also the benchmark: https://github.com/bitcoin/bitcoin/pull/20789#issuecomment-752961937